### DISCUSSION TITLE
Support Renovate Approve Bot for Bitbucket

### DISCUSSION THREAD

**Discussion by maxbrunet** [renovatebot/renovate#12579] [Suggest an Idea]
Hello 👋 

I have written a short script doing just that: https://github.com/maxbrunet/renovate-approve-bot

It does not seem to exist anything else fulfilling this function out there, and I was wondering if you would be open to make it the "official" solution and have it transferred to the @renovatebot org?

Of course, I am open to any feedback or request you would have before considering a transfer.

The tech stack should be familiar to anyone who is used to work with the Renovate ecosystem.
It is very simple, basics (tests, linting, builds, releases) are automated, and I am OK to remain the maintainer for now.

The project would obviously require a new name, and I am not an experienced JS developer, so I would not be surprise if you find oddities in the code.

**Comment by rarkins** [comment]
I think it works a little differently to the main approve bot, right? Could you summarize how it works? e.g. every time it runs, it gets the list of open PRs for the renovate user, and approves them all, including if approved previously and including if they're not configured for automerge? And how do you run it, e.g. do you schedule it every hour, or you run it immediately after Renovate each time, etc?

**Comment by maxbrunet** [comment]
Yes, it is a little different, the Github bot is based on Probot and approves PR via a webhook event, this one runs on a schedule, it will:

1. Get all the PRs from the Renovate Bot user
2. Filter out PRs where auto-merge is not enabled (as it goes through the pagination)
3. Approve the auto-merge PRs (HTTP 200 is approved and HTTP 409 is already approved)

(Filtering if they are already approved requires an extra HTTP call for each PR to get the details and search thought the list of participants if our Approve user has already approved it. I had that logic in a first iteration, but I considered that it was cheaper and simpler to just try to approve anyway)

I run Renovate on a hourly schedule (`0 */1 * * *`) and the Renovate Approve Bot on an hourly schedule on the 50th minute (`50 */1 * * *`), both in Kubernetes Cronjobs. This allows enough time for Renovate to finish before approval starts, and approval to finish before the next run of Renovate.

This works well for self-hosted since they are both scheduled jobs, and compared to a webhook, you do not have to worry about misses. If ever a PR is missed for some reason (or its config changes after being opened), it should be caught on the next run.

A webhook may make more sense for the WhiteSource Renovate which is both schedule and event based (if I am correct), but only Github and Gitlab are supported, not Bitbucket.

**Reply by rarkins** [reply]
Great! I think if you can add a "How it Works" section before "Usage" and capture the above as part of the readme, then we are good to accept it

**Reply by maxbrunet** [reply]
Thank you! I have opened maxbrunet/renovate-approve-bot#42 to add the "How it Works" section, I will wait for your review to merge it.

We also need to pick a name since there is already a [renovatebot/renovate-approve-bot](github.com/renovatebot/renovate-approve-bot), simply `renovate-approve-bot-bitbucket`?

May I ask you if you could invite me to your Slack, please? It might be easier to communicate there to finalize the details of the transfer. You can find my email in the [PR patch](https://patch-diff.githubusercontent.com/raw/maxbrunet/renovate-approve-bot/pull/42.patch).

**Reply by maxbrunet** [reply]
Hello @rarkins, I have merged the above PR, how would you like to proceed for the rest? ^

**Reply by rarkins** [reply]
Assuming you're happy to transfer it in, I can think of these steps:
- It needs to be renamed so that it's clearly for Bitbucket Cloud, maybe needs to be `renovate-approve-bot-bitbucket-cloud`
- You would transfer it to me (@rarkins)
- I can then transfer it to the @renovatebot org

Apart from that I think we leave it as-is

**Reply by maxbrunet** [reply]
Sure, transferring it to you sounds good, a few details I would like to confirm:

- Soon only one Bitbucket will remain, [Server is set for end of support in less than 3 years](https://www.atlassian.com/migration/assess/journey-to-cloud), so is `-cloud` still necessary?
- Would you mind if I remain admin on the repository? At least for some time, I would like to be able to make sure all the automation/integrations are working and fix them in autonomy if needed. Of course, I will go through PR review from the Renovate team for any code change.
- Are you ok with using ghcr.io for container images? I will migrate all existing builds to the new registry

Small changes will be needed to update references to the old repository location, it will probably be my 1st PR

**Reply by rarkins** [reply]
> Sure, transferring it to you sounds good, a few details I would like to confirm:
> 
> 
> 
> - Soon only one Bitbucket will remain, [Server is set for end of support in less than 3 years](https://www.atlassian.com/migration/assess/journey-to-cloud), so is `-cloud` still necessary?

Best to keep cloud because confusion will remain for some time. Also there's Bitbucket Data Centre too.

> - Would you mind if I remain admin on the repository? At least for some time, I would like to be able to make sure all the automation/integrations are working and fix them in autonomy if needed. Of course, I will go through PR review from the Renovate team for any code change.

Yes, please do

> - Are you ok with using ghcr.io for container images? I will migrate all existing builds to the new registry

Sure, sounds good. I think the default GITHUB_TOKEN in actions can hopefully be used for this, so that no additional credentials are necessary?

> 
> 
> Small changes will be needed to update references to the old repository location, it will probably be my 1st PR

**Reply by viceice** [reply]
@rarkins i think he can directly send a Transfer request to renovatebot org. You should then get a notification to accept that.

**Reply by rarkins** [reply]
Sure, we can try that. Pretty sure last time we tried though it didn't work - you could transfer to individuals, or an individual could transfer a repo they own to an org they are admin of.

**Reply by maxbrunet** [reply]
> To transfer a repository that you own to an organization, you must have permission to create a repository in the target organization.

https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository

So if you want me to transfer directly, you would need to give me this permission, I can understand if you do not want that, in which case I'm ok about doing it via @rarkins

**Reply by rarkins** [reply]
Yes, please transfer via me as I don't want to grant anyone admin access to the @renovatebot org

**Reply by maxbrunet** [reply]
I would not need admin access, member would be enough:
https://docs.github.com/en/organizations/managing-peoples-access-to-your-organization-with-roles/roles-in-an-organization#permissions-for-organization-roles

**Reply by rarkins** [reply]
No thanks

**Reply by maxbrunet** [reply]
I have sent out the transfer request. I might need your help to create the ghcr.io repository, not totally sure how it works with organization, once the first artifact is uploaded and it is linked to the repository, it should be easy.

**Reply by viceice** [reply]
the `ghcr.io` repo will created on first push and inherit permissions from repo.

**Reply by maxbrunet** [reply]
As I suspected, I do not have the permission to create it:
```
$ regctl image copy ghcr.io/maxbrunet/renovate-approve-bot:1.0.0 ghcr.io/renovatebot/renovate-approve-bot-bitbucket-cloud:1.0.0
WARN[0004] Failed to push manifest                       err="Failed to put manifest ghcr.io/renovatebot/renovate-approve-bot-bitbucket-cloud@sha256:247bc3687fcace756ff55e2edcb895c27e891c845646a0db5e2d5d06c53f2f74: Unauthorized [http 403]: {\"errors\":[{\"code\":\"DENIED\",\"message\":\"permission_denied: create_package\"}]}\n" target="ghcr.io/renovatebot/renovate-approve-bot-bitbucket-cloud:1.0.0"
Failed to put manifest ghcr.io/renovatebot/renovate-approve-bot-bitbucket-cloud@sha256:247bc3687fcace756ff55e2edcb895c27e891c845646a0db5e2d5d06c53f2f74: Unauthorized [http 403]: {"errors":[{"code":"DENIED","message":"permission_denied: create_package"}]}
```
But hopefully the pipeline should, so merging renovatebot/renovate-approve-bot-bitbucket-cloud#46 should create it, and as you said I should get the permissions via the repo.

**Comment by rarkins** [comment]
Now transferred: https://github.com/renovatebot/renovate-approve-bot-bitbucket-cloud

**Reply by rarkins** [reply]
Thanks @maxbrunet
